<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
  </style>
  <link rel="stylesheet" href="github-pandoc.css" type="text/css" />
</head>
<body>
<big><a href="https://wurmlab.github.io">Wurm lab</a></big>
<h1 id="computing-in-r---practical-3">Computing in R - Practical 3</h1>
<p>BIO309: Research Methods and Communication II</p>
<p>October 2017</p>
<h2 id="introduction">Introduction</h2>
<p>This session will be a recap of what we saw in <a href="https://wurmlab.github.io/SBC361-programming-in-R/practical1.html">practical1</a> and <a href="https://wurmlab.github.io/SBC361-programming-in-R/practical2.html">practical2</a>. You can also use this session to revisit the parts of the last practicals that are still not entirely clear.<br /><strong>REMEMBER</strong> The important thing for the exam is that you understand what you are doing. It is better to understand well a few exercises than finishing all exercises without entirely knowing what is going on.</p>
<h2 id="recap-questions">Recap questions</h2>
<p>To get back into gear in terms of thinking about programming, we’ll start with a quick recap exercise of the material from the last weeks:</p>
<h4 id="intro-q1.-what-would-be-the-outcome-of-the-code-answer---repx-c42-24-times-42-and-what-would-then-be-the-outcome-of-the-code-meananswer">Intro Q1. What would be the outcome of the code <code>answer &lt;- rep(x = c(42, 24), times = 42)</code>? And what would then be the outcome of the code <code>mean(answer)</code>?</h4>
<h4 id="intro-q2.-use-the-preloaded-r-data-set-indometh.-subset-this-data-to-return-only-those-rows-for-which-the-concentration-is-strictly-between-1-and-2.-what-is-the-average-mean-concentration-for-this-subset">Intro Q2. Use the preloaded R data set 'Indometh'. Subset this data to return only those rows for which the concentration is strictly between 1 and 2. What is the average (mean) concentration for this subset?</h4>
<h4 id="intro-q3.-using-regular-expressions-how-would-you-extract-all-the-words-except-antilope-in-the-vector-ccameleopard-eop4a-kiloparsec-antilope-in-r-there-are-a-few-ways-of-getting-a-result.-find-three-ways-to-answer-this-question.">Intro Q3. Using regular expressions, how would you extract all the words except <code>antilope</code> in the vector <code>c(&quot;cameleopard&quot;, &quot;eop4a&quot;, &quot;kiloparsec&quot;, &quot;antilope&quot;)</code>? In R, there are a few ways of getting a result. Find three ways to answer this question.</h4>
<h4 id="intro-q4.-write-a-loop-that-iterates-over-the-numbers-10-to-100-and-stores-this-number-in-a-separate-vector.-can-you-explain-the-sequence-of-events-that-the-computer-goes-through-during-say-the-20th-iteration">Intro Q4. Write a loop that iterates over the numbers 10 to 100 and stores this number in a separate vector. Can you explain the sequence of events that the computer goes through during, say, the 20th iteration?</h4>
<h4 id="intro-q5-a.-write-a-loop-that-iterates-over-the-numbers-16-to-49-and-prints-out-the-square-root-of-the-number-each-time-through-you-may-have-to-search-around-for-the-square-root-function.-yes-we-absolutely-want-you-to-write-a-loop-to-do-this.">Intro Q5-A. Write a loop that iterates over the numbers 16 to 49 and prints out the square root of the number each time through (you may have to search around for the square root function). Yes we <em>absolutely</em> want you to write a loop to do this.</h4>
<h4 id="intro-q5-b.-make-the-loop-from-intro-q5-a-store-the-results-to-a-separate-vector-called-my_square_roots-instead-of-just-printing-the-results.-whats-the-value-of-the-3rd-iteration-whats-the-sum-of-the-square-roots-of-the-numbers-16-to-49">Intro Q5-B. Make the loop from Intro Q5-A store the results to a separate vector called <code>my_square_roots</code> instead of just printing the results. What's the value of the 3rd iteration? What's the sum of the square roots of the numbers 16 to 49?</h4>
<h4 id="intro-q6.-write-a-loop-that-iterates-over-all-even-numbers-between-30-and-90.-at-each-iteration-run-your-function-to-convert-each-of-these-even-numbers-from-kilometres-into-miles-you-created-this-function-in-q2-from-practical2.-the-loop-should-store-the-results-in-a-separate-vector.">Intro Q6. Write a loop that iterates over all even numbers between 30 and 90. At each iteration, run your function to convert each of these even numbers from kilometres into miles (you created this function in Q2 from <a href="https://wurmlab.github.io/SBC361-programming-in-R/practical2.html">practical2</a>). The loop should store the results in a separate vector.</h4>
<h4 id="intro-q7.-numbers-in-fibonaccis-sequence-are-characterized-by-the-fact-that-every-number-after-the-first-two-is-the-sum-of-the-two-preceding-onese.g.-1-1-2-3-5....-this-seemingly-simple-sequence-can-be-used-to-explain-different-biological-patterns-ranging-from-the-arangement-of-leaves-on-a-stem-or-the-flowering-of-an-artichoke-to-the-patterns-of-inheritance-of-x-chromosomes.-write-a-for-loop-that-calculates-the-first-20-numbers-in-a-fibonacci-sequence-strating-from-1.">Intro Q7. Numbers in Fibonacci's sequence are characterized by the fact that every number after the first two is the sum of the two preceding ones(e.g., 1, 1, 2, 3, 5,...). This seemingly simple sequence can be used to explain different biological patterns ranging from the arangement of leaves on a stem or the flowering of an artichoke to the patterns of inheritance of X chromosomes. Write a <code>for</code> loop that calculates the first 20 numbers in a Fibonacci sequence strating from 1.</h4>
<p><em>Hint - you will need to create an empty vector to store every element in your Fibonacci sequence</em></p>
<h4 id="intro-q8.-create-a-function-that-can-run-this-loop-to-generate-a-fibonacci-sequence-of-any-size-starting-from-any-given-number.-so-for-example-a-sequence-of-length-x-from-number-y-where-x-and-y-are-inputs-to-the-function.">Intro Q8. Create a function that can run this loop to generate a Fibonacci sequence of any size, starting from any given number. So for example, a sequence of length <code>x</code> from number <code>y</code>, where <code>x</code> and <code>y</code> are inputs to the function.</h4>
<h2 id="palindromic-sequences">Palindromic Sequences</h2>
<p>Palindromes are arrangements of words or letters which read the same way whether you read them backwards or forwards, such as the phrase ‘Never odd or even’. In molecular biology, many restriction enzyme sites are palindromic.</p>
<h4 id="q1.-write-a-function-that-assesses-whether-a-given-word-or-phrase-is-a-palindrome.">Q1. Write a function that assesses whether a given word or phrase is a palindrome.</h4>
<p>Before starting to code, think about the steps that you would need to go through in order to judge if something is a palindrome or not. Write these steps as comments in an R script window, then think about how you can tell the computer to execute those steps. Only start writing the code when you have a plan of what you want to do. Don’t be afraid to test lines independently in the console and to use easy test cases where you know the answer in order to check that your function works.</p>
<p>The bottom section of the R help sheets normally have examples of how a command can be used. Sometimes one of these examples will be a way to solve the problem that you are currently working on. The <code>strsplit</code> helpsheet is particularly interesting in relation to this question.</p>
<h4 id="q2.-palindromes-in-molecular-biology-work-a-little-differently-in-that-the-sequence-is-the-same-if-read-5-to-3-on-one-strand-or-5-to-3-on-the-complementary-strand.-for-example-given-the-sequence-acctaggt-its-complement-would-be">Q2. Palindromes in molecular biology work a little differently, in that the sequence is the same if read 5’ to 3’ on one strand or 5’ to 3’ on the complementary strand. For example, given the sequence ACCTAGGT, its complement would be:</h4>
<h4 id="acctaggt-3">5’ ACCTAGGT 3’</h4>
<h4 id="section">||||||||</h4>
<h4 id="tggatcca-5">3’ TGGATCCA 5’</h4>
<h4 id="both-acctaggt-and-its-complementary-sequence-tggatcca-are-identical-if-read-from-5-to-3-that-is-acctaggt-is-a-palindromic-sequence.-write-a-function-that-assesses-whether-a-dna-sequence-is-palindrome.-you-will-need-to-convert-it-to-its-reverse-complement-and-then-compare-this-to-the-original-sequence.-again-there-are-commands-which-may-help-you-in-the-working-with-dna-data-section-from-the-last-session.">Both ACCTAGGT and its complementary sequence TGGATCCA are identical if read from 5’ to 3’, that is, ACCTAGGT is a palindromic sequence. Write a function that assesses whether a DNA sequence is palindrome. You will need to convert it to its reverse complement and then compare this to the original sequence. Again, there are commands which may help you in the ‘Working with DNA data’ section from the <a href="https://wurmlab.github.io/SBC361-programming-in-R/practical2.html">last session</a>.</h4>
<h2 id="open-reading-frames">Open Reading Frames</h2>
<p>Protein-coding regions in the genome can be predicted by detecting open reading frames. An open reading frame normally begins with the start codon ‘ATG’ and ends at one of three possible stop codons, ‘TGA’, ‘TAA’ and ‘TAG’. The sequence in between these two points is arranged in 3-base codons.</p>
<h4 id="q3.-write-a-function-which-uses-regular-expressions-to-detect-if-a-given-sequence-contains-an-open-reading-frame.-test-it-on-the-following-sequences">Q3. Write a function which uses regular expressions to detect if a given sequence contains an open reading frame. Test it on the following sequences:</h4>
<p><em>Hint - an open reading frame should always start with the start codon 'ATG' and end with one of the stop codons. Additionally, the length of the entire sequence should be a multiple of 3. If any of these features are not met, the sequence is not a functional open reading frame</em></p>
<pre><code>ATGGATTTTTAG
ATGGATTTTCTAG
CTAATGGATTTTTGAAT
atgctaaactaa
TCGATTAA</code></pre>
<h4 id="hacker-q4a.-instead-of-having-the-function-return-whether-your-sequence-contains-an-open-reading-frame-have-the-function-return-a-string-containing-the-open-reading-frame-if-there-is-one-and-null-if-there-is-not-you-will-need-to-look-up-if-statements.">Hacker Q4a. Instead of having the function return whether your sequence contains an open reading frame, have the function return a string containing the open reading frame if there is one and NULL if there is not (you will need to look up &quot;if statements&quot;).</h4>
<h4 id="hacker-q4b.-this-time-check-both-the-forward-and-reverse-strand-for-open-reading-frames.-return-forward_only-if-there-is-an-open-reading-frame-only-on-the-forward-strand-reverse_only-if-only-on-the-reverse-both-if-in-both-or-none-if-in-neither.">Hacker Q4b. This time, check both the forward and reverse strand for open reading frames. Return <code>&quot;forward_only&quot;</code> if there is an open reading frame only on the forward strand, <code>&quot;reverse_only&quot;</code> if only on the reverse, <code>&quot;both&quot;</code> if in both or &quot;none&quot; if in neither.</h4>
<h2 id="species-names">Species names</h2>
<p>Run the following line of code to import the <code>butterfly_sample</code> and the <code>butterfly_reference</code> data frames:</p>
<pre class="sourceCode R"><code class="sourceCode r">butterfly_sam_url &lt;-<span class="st"> &quot;http://wurmlab.github.io/SBC361-programming-in-R/butterfly_sample.csv&quot;</span>
butterfly_sample  &lt;-<span class="st"> </span><span class="kw">read.csv</span>(butterfly_sam_url, <span class="dt">header =</span> <span class="ot">TRUE</span>)

butterfly_ref_url   &lt;-<span class="st"> &quot;http://wurmlab.github.io/SBC361-programming-in-R/butterfly_reference.csv&quot;</span>
butterfly_reference &lt;-<span class="st"> </span><span class="kw">read.csv</span>(butterfly_ref_url, <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre>
<p>The <code>butterfly_reference</code> data frame contains the species name and the common name of a number of butterflies.</p>
<p>The <code>butterfly_sample</code> data frame contains information on butterflies caught in sweep netting surveys in two locations under different pesticide treatments (locations A and B). This data was collected by multiple people, who have recorded the common names of the species they encountered (without using a standard letter case). In order to be able to compare the diversity between the two different sites, you will need to standardise the names.</p>
<h4 id="q5-write-a-function-that-adds-a-new-column-in-the-data-frame-that-contains-the-correct-latin-species-name-for-each-record-in-the-butterfly_sample.">Q5 Write a function that adds a new column in the data frame that contains the correct Latin species name for each record in the <code>butterfly_sample</code>.</h4>
<p>TIP: There are several ways to do this. Remember that R is case sensitive, so you will need to account for case differences in your function. <code>grep</code> and <code>gsub</code> both allow you to set an <code>ignore.case = TRUE</code> option. Alternately, you could use the R commands <code>toupper()</code> and <code>tolower()</code>. Use the help pages to see how these work, which you can access by typing a question mark before the command - <code>?toupper</code>.</p>
<h4 id="q6.-from-the-data-above-which-location-has-the-greatest-number-of-different-species">Q6. From the data above, which location has the greatest number of different species?</h4>
<h4 id="q7.-which-genus-has-been-caught-the-greatest-number-of-times">Q7. Which genus has been caught the greatest number of times?</h4>
<h1 id="bonus-hacker-question">Bonus hacker question</h1>
<p>This question is an extension of the question Q5. Again, we will analyse a data frame containing the species of butterflies observed in two locations (A and B). This time, however, some of the people who did the sampling recorded the common name of the species, while others recorded the latin name (all with inconsistent letter case).</p>
<h4 id="q8-write-a-script-that-adds-a-new-column-in-the-data-frame-that-contains-the-correct-latin-species-name-for-each-record-in-the-data-frame-imported-below.">Q8 Write a script that adds a new column in the data frame that contains the correct Latin species name for each record in the data frame imported below.</h4>
<pre class="sourceCode R"><code class="sourceCode r">butterfly_sam_bonus_url &lt;-<span class="st"> &quot;http://wurmlab.github.io/SBC361-programming-in-R/butterfly_sample_bonus.csv&quot;</span>
butterfly_sample_bonus  &lt;-<span class="st"> </span><span class="kw">read.csv</span>(butterfly_sam_url, <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre>
<p>One of the difficulties of this exercise is that you will have to perform a different process depending on whether the sample already has its latin name or not. You may find that using <code>if</code> statements may be helpful (which you will have to look up). Alternatively, you may want to subset the data into the two groups (where the transformation is either from latin to latin or common to latin), and do the transformation independently on each.</p>
<p>This question is not easy! But it is typical of the sort of thing researchers do from day to day and a very good test of all the things you have learned this year!</p>
<hr>
<small>Copyright 2017 <a href=https://github.com/wurmlab/genomicscourse/graphs/contributors>Authors</a>. All rights reserverd.

<!-- Global Site Tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-471841-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments)};
  gtag('js', new Date());

  gtag('config', 'UA-471841-2');
</script>
</body>
</html>
